uboot: init (firmwareOdroidC2/C4, ubootOdroidC4) Hardkernel's Odroid C4 and Odroid C2 board support, drop Armbian dependency

arapov opened the pull request:
###### Motivation for this change

This enables Odroid C4 board in NixOS with Linux kernel version 5.8+.
Also retires dependency on 3rd-party repository from Armbian.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested Odroid C2/C4 actually boots from produced u-boot.
- [x] Odroid C4 boot from SD tested.
- [x] Odroid C4 boot from eMMC tested.
- [x] Odroid C2 boot from SD tested.
- [x] Odroid C2 boot from eMMC tested.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

arapov committed:
uboot: init (firmwareOdroidC2/C4) hardkernel's Odroid C2/C4 fw blobs

This patch brings Hardkernel Odroid C2 and Odroid C4 firmware.
- https://wiki.odroid.com/odroid-c2/software/building_u-boot
- https://wiki.odroid.com/odroid-c4/software/building_u-boot

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov committed:
uboot: init (ubootOdroidC4) Hardkernel's Odroid C4 board support

Odroid C4: Cortex-A55 CPU, Mali-G31 GPU.
More data: https://wiki.odroid.com/odroid-c4/

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov committed:
maintainers: add aarapov

ssia

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov commented:
fyi, booting from eMMC on OdroidC2 is failing for me in u-boot v2020.07 and v2020.10. Though works flawlessly in v2020.04. I am triaging it and working on a fix right now.

arapov commented:
> fyi, booting from eMMC on OdroidC2 is failing for me in u-boot v2020.07 and v2020.10. Though works flawlessly in v2020.04. I am triaging it and working on a fix right now.

this is the commit which breaks eMMC for OdroidC2:
https://gitlab.denx.de/u-boot/u-boot/-/commit/dd5f2351e99aad8fcbedbc1305b8b51b09952336

looking into...

arapov commented:
> > fyi, booting from eMMC on OdroidC2 is failing for me in u-boot v2020.07 and v2020.10. Though works flawlessly in v2020.04. I am triaging it and working on a fix right now.
> 
> this is the commit which breaks eMMC for OdroidC2:
> https://gitlab.denx.de/u-boot/u-boot/-/commit/dd5f2351e99aad8fcbedbc1305b8b51b09952336
> 
> looking into...

This is the line from the patch that breaks eMMC: 
https://gitlab.denx.de/u-boot/u-boot/-/blob/master/arch/arm/dts/meson-gxbb.dtsi#L411

I will send out patch to respective upstreams.

lopsided98 commented on pkgs/misc/uboot/default.nix:
```diff
@@ -179,50 +181,113 @@ in {
-      cat '${firmwareBlobs}/gxb/bl2.package' fip.bin > boot_new.bin
-      '${buildPackages.meson-tools}/bin/amlbootsig' boot_new.bin u-boot.img
-      dd if=u-boot.img of=u-boot.gxbb bs=512 skip=96
+      ${firmwareOdroidC2}/fip_create --dump fip.bin
```
`fip_create` is distributed as a binary blob for x86_64 only, so we should avoid it.

arapov replied:
reverted back to fiptool

lopsided98 commented on pkgs/misc/uboot/default.nix:
```diff
@@ -179,50 +181,113 @@ in {
+      ${firmwareOdroidC2}/fip_create --dump fip.bin
+
+      cat ${firmwareOdroidC2}/bl2.package fip.bin > boot_new.bin
+      ${firmwareOdroidC2}/aml_encrypt_gxb --bootsig \
```
Same here. `amlbootsig` is an open source re-implementation of this tool.

arapov replied:
reverted back to amlbootsig

lopsided98 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,126 @@
+      "sd_fuse/sd_fusing.sh" "sd_fuse/bl1.bin.hardkernel"
+    ];
+
+    extraMeta.platforms = [ "aarch64-linux" ];
```
For this to be accurate, we need to not install the x86_64 binaries (which aren't needed as we have open source alternatives).

arapov replied:
documented this case in code and made builds gracefull

arapov replied:
Fixed, this derivative does not produce x86_64 binaries.

lopsided98 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,126 @@
+      "fip/g12a/aml_encrypt_g12a" "sd_fuse/sd_fusing.sh"
+    ];
+
+    extraMeta.platforms = [ "aarch64-linux" ];
```
I think `x86_64-linux` makes more sense here because all the precompiled Linux binaries are for x86_64. Then you can use it through `buildPackages`. Also, add a comment explaining this unusual sitation.

arapov replied:
ditto

lopsided98 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,126 @@
+      '';
+
+      makeFlags = [
+        "CROSS_COMPILE=${stdenv.cc.targetPrefix}"
```
Does this work for native builds?

arapov replied:
From what I see, it does work.

lopsided98 commented on pkgs/misc/uboot/default.nix:
```diff
@@ -179,50 +181,113 @@ in {
+              ${firmwareOdroidC4}/acs.bin bl21_zero.bin \
+              bl2_new.bin bl2
+
+      ${firmwareOdroidC4}/aml_encrypt_g12a --bl30sig --input bl30_new.bin \
```
Ideally we would have an open source alternative to this tool, but I can't find any. This means native builds will not work.

arapov replied:
I will keep watching for it.

arapov commented:
FYI, discussion for the possible eMMC fix: https://groups.io/g/u-boot-amlogic/topic/meson_gxbb_regression_since/77816805
@lopsided98, thanks for comments, I should have a time next week to address all the notes.

samueldr commented on pkgs/misc/uboot/default.nix:
```diff
@@ -179,27 +181,21 @@ in {
+    # hack: make board able to boot from eMMC
+    #     : wip https://groups.io/g/u-boot-amlogic/topic/meson_gxbb_regression_since/77816805
+    preBuild = ''
+      sed -i "388d" arch/arm/dts/meson-gxbb.dtsi
```
Can we verify the line content before deleting it, for whenever a future u-boot update changes the device tree sources, it'd be best to break in a controlled manner rather than delete an arbitrary line.

arapov replied:
Yeah, very true. I switched this to *.patch with the relevant comments.

arapov commented:
I decided to keep the odroid-c4 hack in git history, so that it is more clear what has actually happened in mmc driver for gx/sm1 boards.

lopsided98 commented on pkgs/misc/uboot/default.nix:
```diff
@@ -212,17 +206,116 @@ in {
+    # to SD or eMMC card.
+    # Due to absence of the "aml_encrypt_g12a" signing tool binary built for
+    # aarch64 - Odroid C4 must be cross-compiled on x86_64 machine.
+    dontBuild = stdenv.buildPlatform.isAarch64;
```
It seems like it would be less confusing to just add an assertion (using `assertMsg`) explaining that it must be cross-compiled.

arapov committed:
uboot: (ubootOdroidC2) retire Armbian dependency

Retire dependency on Armbian provided board blobs.
Use official hardkernel's ones instead.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov commented:
All the initial hacks were fixed and fixes are cherry-picked. Fixes are in staging for 2021.01 release.

sbruder commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,129 @@
+    src = fetchgit {
+      url = "https://github.com/hardkernel/u-boot.git";
+      rev = "12c58e94e533b85a19d2f83d1c0a34345764ca07"; # "odroidg12-v2015.01"
+      sha256 = "06xmj8ksyx21qfdyq84lj029778vj30kwlrb91yj7bhwjxcz0qzn";
```
```suggestion
      sha256 = "110f8i2aj8p8b6ycw2hvilkp47vw754vdwv6vkpap8wmgkk90fjz";
```

That hash seems to be wrong:

    hash mismatch in fixed-output derivation '/nix/store/9mlfg5xlydnncrkin0l13vplf36ism82-u-boot-12c58e9':
      wanted: sha256:06xmj8ksyx21qfdyq84lj029778vj30kwlrb91yj7bhwjxcz0qzn
      got:    sha256:110f8i2aj8p8b6ycw2hvilkp47vw754vdwv6vkpap8wmgkk90fjz

arapov replied:
Hmm... It is hardkernel's repo got a new commit and I thought I pinned it with `rev = ...;`.

sbruder reviewed (changes requested):
After changing the u-boot source hash (see inline comment) and erasing petitboot from the SPI flash as described [here](https://www.armbian.com/odroid-hc4/), this successfully booted on my Odroid HC4 (compatible with Odroid C4).

Thank you!

arapov committed:
uboot: (firmwareOdroidC2/C4) drop patches for 2021.01

u-boot 2021.01 was released, required patches to run hardkernel's boards
are no longer needed as they made the release.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov committed:
uboot: (firmwareOdroidC2/C4) switch fetchgit to fetchFromGitHub

This solves repo pinpointing to particular commit number, thus prevents
failing builds when the repos are updated.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov commented:
@lopsided98, could you check this patchset please. I will be happy to see this merged eventually. :-)

lopsided98 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,142 @@
+
+    # https://wiki.odroid.com/odroid-c2/software/building_u-boot
+    prePatch = ''
+      patch -p1 < ${fetchpatch {
```
I guess this was done to match the instructions given in that link, which tell you to use `git merge`, but it might be clearer and easier to maintain (e.g. you wouldn't need to add more patches if hardkernel ever decides to update the firmware) if you were to fetch the `u-boot_firmware` repo and just copy the firmware directories into place in the u-boot tree.

arapov replied:
This is almost the way it was before the last commit. I did think about just making a copy over, though I decided to keep a little more control here. I don't anticipate there will be any other change as the last one is dated to Feb 2017. And if there will be I do want to know what it is.
That said it is not an issue to maintain it, and in case we need to add a patch - we have a control in a way we don't get anything creepy and this new change won't break anything unexpectedly.

lopsided98 replied:
ok, fine with me

lopsided98 commented on pkgs/misc/uboot/default.nix:
```diff
@@ -217,17 +206,95 @@ in {
+
+  ubootOdroidC4 = let in
+  assert
+    (lib.assertMsg (!stdenv.buildPlatform.isAarch64)
```
```suggestion
    (lib.assertMsg stdenv.buildPlatform.isx86_64
```

arapov replied:
Makes perfect sense.

lopsided98 reviewed (changes requested):
I successfully tested this with my ODROID-C2. This looks good except for the two suggestions.

arapov commented:
@lopsided98, thanks for the ACK. Could you point me to a document that describes the process of getting this code merged? Or explain it here briefly. I am new to nixos development model, need to close some gaps.

lopsided98 commented:
You normally need to find someone with commit privileges who is willing to review and approve the PR. That varies in difficulty depending on what part of nixpkgs you are working on. In this case, it will probably be @samueldr, who is the primary maintainer for U-Boot.

arapov commented:
ping @samueldr, we need your attention here.

arapov commented:
One more *ping* to @samueldr, before I try to reach out you elsewhere... :)

arapov commented:
@lopsided98, is there any specific way to reach out to @samueldr? I tried irc, e-mail, github, whatnot and the guy is not responsive. :-(

arapov committed:
meson64-tools: init

an open alternative to aml_encrypt_g12a and aml_encrypt_g12b.
Proprietary tool is not available for aarch64 architecture, which is a
show stopper for building u-boot images for the boards like Odroid C4.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov committed:
uboot: (ubootOdroidC4) use meson64-tools instead of aml_encrypt_g12a

This makes aarch64 builds possible.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov commented:
@lopsided98, it turned out meson64-tools just flawlessly does the work of aml_encrypt_g12a. I tested builds on aarch64 and crossbuild on x86_64, as well as booting the board.

lopsided98 commented on pkgs/misc/meson64-tools/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+stdenv.mkDerivation rec {
+  pname = "meson64-tools";
+  version = "0.0";
```
This should be some form of date based version, although the exact format is subject to some debate: https://github.com/NixOS/nixpkgs/pull/100833. I'd suggest `unstable-YYYY-MM-DD` since that is the most common and it will just become part of the eventual treewide change.

arapov replied:
Yeah, I was curious about it when I was typing "0.0". Thanks

lopsided98 commented:
samueldr is quite active on #nixos-aarch64. Also, he mentioned this PR a few days ago, see: https://logs.nix.samueldr.com/nixos-aarch64/2021-03-13#4730884;

arapov commented:
> samueldr is quite active on #nixos-aarch64. Also, he mentioned this PR a few days ago, see: https://logs.nix.samueldr.com/nixos-aarch64/2021-03-13#4730884;

Thanks, I overlooked #nixos-aarch64 channel. @samueldr is indeed active there. :)

SuperSandro2000 commented on pkgs/misc/meson64-tools/default.nix:
```diff
@@ -0,0 +1,27 @@
+    sha256 = "1487cr7sv34yry8f0chaj6s2g3736dzq0aqw239ahdy30yg7hb2v";
+  };
+
+  buildInputs = with buildPackages; [ openssl bison yacc flex bc python3 python2 ];
```
Please don't mix python2 and python3.

arapov replied:
@SuperSandro2000, I did simple `s/python2/python/g` here. python2 roots from mbedtls dependency, and there are no bits of python that wouldn't work in python3. tested (tm).

SuperSandro2000 replied:
But why do we need python2 then?

arapov replied:
@SuperSandro2000, we don't, a script that runs tests in meson64-tools wants a patch. I will suggest it to the relevant repo later.

SuperSandro2000 commented on pkgs/misc/meson64-tools/default.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with lib; {
+    homepage = "https://github.com/angerman/meson64-tools";
+    description = "Tools for Amlogic Meson ARM64 platforms";
+    maintainers = with maintainers; [ aarapov ];
```
Missing license.

arapov replied:
There is no license specified. Relevant [issue](https://github.com/angerman/meson64-tools/issues/2) in meson64-tools repository is opened. As long as we don't know - it is **unfree**.

sbruder reviewed (approved):
I don’t know enough to review the code, but the current version builds successfully and boots without problems on my Odroid HC4.

arapov commented:
> I don’t know enough to review the code, but the current version builds successfully and boots without problems on my Odroid HC4.

Thanks, good to know. I don't have any HC4 around. :)

SuperSandro2000 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,136 @@
+    };
+
+    # https://wiki.odroid.com/odroid-c2/software/building_u-boot
+    prePatch = ''
```
Why are the patches not in patches/patchPhase?

arapov replied:
AFAIK, we have to use either prePatch or postPatch phase here so that we can explicitly call patch util to patch the sources (instead of using builtin patch func, which is not capable to move and create new files).

@SuperSandro2000, I also might not be aware of some other elegant way to apply this patches here.

SuperSandro2000 replied:
Then postPatch should be used because prePatch breaks patches put in patches.

lopsided98 replied:
Have you tried using `patches = [ ... ]`? I don't think it has any limitations regarding moving/creating files because it ends up calling the patch command in pretty much the same way you are, see: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L948

arapov replied:
@lopsided98, indeed,... I do remember it didn't work for me back in October and don't remember why. Now it just works. I guess I used it somehow differently.
@SuperSandro2000, even after switching to `patches = [ ... ]` there are few edits in Makefiles. I feel like preBuild would be a better place to make changes in Makefiles, let me know if you think otherwise.

SuperSandro2000 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,136 @@
+    '';
+
+    filesToInstall = [
+      "build/scp_task/bl301.bin" "fip/gxb/bl30.bin" "fip/gxb/bl2.package"
+      "sd_fuse/sd_fusing.sh" "sd_fuse/bl1.bin.hardkernel"
```
```suggestion
      "build/scp_task/bl301.bin"
      "fip/gxb/bl30.bin"
      "fip/gxb/bl2.package"
      "sd_fuse/sd_fusing.sh"
      "sd_fuse/bl1.bin.hardkernel"
```

SuperSandro2000 commented on pkgs/misc/uboot/hardkernel-firmware.nix:
```diff
@@ -0,0 +1,136 @@
+    '';
+
+    filesToInstall = [
+      "build/board/hardkernel/odroidc4/firmware/acs.bin" "build/scp_task/bl301.bin"
+      "fip/g12a/bl2.bin" "fip/g12a/bl30.bin" "fip/g12a/bl31.img"
+      "fip/g12a/ddr3_1d.fw" "fip/g12a/ddr4_1d.fw" "fip/g12a/ddr4_2d.fw"
+      "fip/g12a/lpddr3_1d.fw" "fip/g12a/lpddr4_1d.fw" "fip/g12a/lpddr4_2d.fw"
+      "fip/g12a/diag_lpddr4.fw" "fip/g12a/piei.fw" "fip/g12a/aml_ddr.fw"
+      "sd_fuse/sd_fusing.sh"
```
```suggestion
      "build/board/hardkernel/odroidc4/firmware/acs.bin"
      "build/scp_task/bl301.bin"
      "fip/g12a/bl2.bin"
      "fip/g12a/bl30.bin"
      "fip/g12a/bl31.img"
      "fip/g12a/ddr3_1d.fw"
      "fip/g12a/ddr4_1d.fw"
      "fip/g12a/ddr4_2d.fw"
      "fip/g12a/lpddr3_1d.fw"
      "fip/g12a/lpddr4_1d.fw"
      "fip/g12a/lpddr4_2d.fw"
      "fip/g12a/diag_lpddr4.fw"
      "fip/g12a/piei.fw"
      "fip/g12a/aml_ddr.fw"
      "sd_fuse/sd_fusing.sh"
```

arapov replied:
@SuperSandro2000, I don't mind this change. However interested - why exactly this is better than to keep more than one file referenced on the same line in this particular example?

SuperSandro2000 replied:
It can easily be sorted alphabetically and removing/adding do entries less merge conflicts.

arapov replied:
Makes sense. Fixed.

arapov committed:
uboot: (firmwareOdroidC2/C4) oneline and sort list of files to install

as per review request

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov committed:
uboot: (firmwareOdroidC2/C4) use postPatch instead of prePatch

postPatch should be used because prePatch breaks patches put in patches.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

arapov committed:
uboot: (firmwareOdroidC2/C4) don't invoke patch tool, use patches = [] instead

https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L948
this can do it nicely.

Signed-off-by: Anton Arapov <anton@deadbeef.mx>

SuperSandro2000 commented:
I don't really know kernels in nixpkgs I must admit.

samueldr commented:
Started looking into this for real

 - https://logs.nix.samueldr.com/nixos-aarch64/2021-04-12#1618248423-1618256207;

Just wanted to state it here.

samueldr commented:
Here's the strict minimum implementing what I had in mind.

 - https://github.com/samueldr/nixpkgs/commits/PR101454

I will build on top of this for the [`AML-S805X-AC`](https://libre.computer/products/boards/aml-s805x-ac/) board, in a further branch/PR.

@arapov please try with the changed firmware blobs. If it does not work, we can undo the change where I remove the vendor blobs from Hardkernel and use them.

Note that this removes the addition of `sd_fusing.sh` scripts. They did not really work, as it is expected the user's CWD is the directory the files are in. And it's a bit too opinionated in forcing sudo and `sync` calls.

Instead, the user would use the appropriate command listed in the README.md file. Only the C2 is a bit more "complex", but none of the two boards have entirely surprising commands to flash the firmware, compared to other board. It amounts in either one or two `dd` invocations, nothing more.

arapov commented:
@samueldr , yeah I will give your branch a test and soaking time. thanks!

samueldr commented:
I force-pushed the minor fixes to the branch. For transparency, one of the builder functions didn't install the README.md file, and I ended up not deleting the python and sh helpers from the firmware files, as they do things not done with the other tools.

I also opened #119371 which will build on top of those changes.

* * *

I am inclined to think the changes are correct and desirable in Nixpkgs. I would simply append those commits on this PR, without squashing anything currently observable. The whole trail is AFAIK buildable. At least for my changes. It clearly shows how it went from the original builds to the final builds.

Any other U-Boot maintainers (e.g. @lopsided98) have opinions here?

Note that it is my understanding that we probably won't really be able to replace those early init blob ever. Though BL31 *may*, at some point, we'll see.

arapov commented:
@samueldr, ssd in my laptop has died. waiting for a replacement, till then I'm not able to build/test the u-boot images.
I tried to build it on my Mac and it fails, and I'm not sure I want to look into it. :-)

erdnaxe commented:
I am able to boot my Odroid HC4 with this PR, it works perfectly!

My Odroid HC4 is now running NixOS 21.04 using <https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/sd-card/sd-image-aarch64-new-kernel-installer.nix> and I removed Petitboot (instructions on https://www.armbian.com/odroid-hc4/).

Patagonicus commented:
Can confirm that this works on the HC4. For anyone trying to reproduce this:

1. Flash https://hydra.nixos.org/build/151751284 to an sdcard.
2. Check out this pull request, currently at 7cf5a0efb9d92991c9cbcdfdbb5349225dcdbbc6, `nix-build . -A pkgsCross.aarch64-multiplatform.ubootOdroidC4`, then `./result/sd_fusing.sh $SDCARDDEV`.
3. Hold down the button on the underside of the HC4 while you plug in the power, which will bypass the petitboot boot loader in the SPI flash and instead load the u-boot from the Micro-SD card.

Once you've confirmed that NixOS boots successfully and you won't need Ubuntu anymore, you can also erase petitboot, which remove the need for pressing the button, see https://nixos.wiki/wiki/NixOS_on_ARM/ODROID-HC4#Petitboot_removal.

Unfortunately with that image I don't get any output on HDMI or UART, but I was able to use the Hardkernel Ubuntu image to rebuild the image with a password for nixos set, which allowed SSH login: check out 8a2ec31e224de9461390cdd03e5e0b0290cdad0b, modify `nixos/modules/installer/sd-card/sd-image-aarch64-new-kernel-installer.nix` to set the password, then build via `nix-build nixos -I nixos-config=nixos/modules/installer/sd-card/sd-image-aarch64-new-kernel-installer.nix -A config.system.build.sdImage`.

Currently trying to see if the Hardkernel kernel will make HDMI and/or UART work. Basing my kernel package off of https://github.com/considerate/nixos-odroidhc4/tree/983c146bc7d1d05594e04f99e808b0a661ed80bd/overlays/kernel for that. That should also hopefully make SATA work (didn't test it with a drive, but at least dmesg and lspci do not show anything with mainline kernel; https://nixos.wiki/wiki/NixOS_on_ARM/ODROID-HC4 also notes that SATA is not working).

erdnaxe commented:
> Unfortunately with that image I don't get any output on HDMI

In my case I was able to get HDMI working with a standard 1080p screen, following your method.

SATA is not working on my side (lspci output is empty).

considerate commented:
> SATA is not working on my side (lspci output is empty).

SATA works for me if I update to a more recent version of uboot and use the `odroid-hc4_defconfig` file which includes `SCSI=y` and `SATA=y`.

https://github.com/u-boot/u-boot/blob/v2022.04/configs/odroid-hc4_defconfig

dnr commented:
Another report: this works great on my HC4. Thanks to everyone who worked on this.

I built u-boot with the instructions [here](https://nixos.wiki/wiki/NixOS_on_ARM/ODROID-HC4), which refer to samueldr's repo, but it looks like it's the same code rebased.

Using nixos 22.05 with stock kernel, currently 5.15.49. SATA works. I have not verified HDMI or UART.

considerate commented:
> I have not verified HDMI or UART.

I have verified that both HDMI and UART works. I'm using the configuration here https://github.com/considerate/nixos-odroidhc4/tree/0c1716bf1d8c7f99ba6ad0334e81240afdc27cbf which is based on this PR.

arapov commented:
> There's some overlap with #245100 that has no merge conflicts currently. So maybe we can get that in and update this one.

I was under impression that this patch is "no-go" because of the https://github.com/Tow-Boot/Tow-Boot project. 

In case it is not true - I can update this PR.

lopsided98 commented:
I don't think Tow-Boot should have any impact on U-Boot packaging in nixpkgs. Was this discussed somewhere?

Binary blobs are never nice, but we already support the ODROID-C2, and the C4 is basically the same. This PR also gets rid of some of the existing C2 blobs.

My only suggestion after reading through this PR again is that perhaps the package attributes should be `firmware-odroid-c2/c4`. Our U-Boot packages use camel case, and I'm guilty of copying it with `armTrustedFirmware`, but this is not the standard convention in nixpkgs.
